### PR TITLE
serialize without id-arena fork

### PR DESCRIPTION
**Pull request by ydnar**
Updates `serde` support (serialization only) for `Resolve` and child types to work without a fork of the `id-arena` crate. It leans heavily on `#[serde(serialize_with = "...")]` to do the heavy lifting for foreign types.

The JSON serialization is opinionated:

1. Enums are flattened where possible, and variants are lowercased, to match values expressed in WIT versus the Rust implementation.
1. Function results are always expressed as a JSON array, with an optional `name` key. This is an affordance to parsers in languages with a less sophisticated type system (e.g. Go). `Anon` results will be a single-item JSON array, e.g.: `[{"type": <type>}]`
2. `ArenaId<T>` are serialized to JSON numbers as indexes into the arrays of items in the outer `Resolve` struct. The IDs are specific to this serialization, and are not globally unique.

### CODE REVIEW

File: crates/wit-component/src/decoding.rs [hunk 1]
```diff
@@ -871,7 +871,7 @@ impl WitPackageDecoder<'_> {
                     .context("failed to convert anonymous result type")?,
             )
         } else {
-            Results::Named(
+            Results::Named(Params(
```
Comment (Mossaka) [inline crates/wit-component/src/decoding.rs]: Could implement `Into` and `From` trait so don't need to wrap a `Params` outside but can call `into()`.

File: crates/wit-parser/Cargo.toml [hunk 2]
```diff
@@ -13,7 +13,7 @@ Tooling for parsing `*.wit` files and working with their contents.
 """
 
 [dependencies]
-id-arena = { git = "https://github.com/Mossaka/id-arena-fork", rev = "cae2a0a63d505adec402289eea90c52d3205007a" }
+id-arena = "2"
```
Comment (Mossaka) [inline crates/wit-parser/Cargo.toml:16]: nice!

File: crates/wit-parser/src/id_arena_.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+impl<T> Id<T> {
+    /// Get the index within the arena that this id refers to.
+    #[inline]
+    pub fn index(&self) -> usize {
+        self.idx
+    }
+}
+
+/// Our `ArenaBehavior` implementation.
```
Comment (Mossaka) [inline crates/wit-parser/src/id_arena_.rs]: Provide a better docs

File: crates/wit-parser/src/lib.rs [hunk 4]
```diff
@@ -562,7 +583,24 @@ pub struct Docs {
     pub contents: Option<String>,
 }
 
-pub type Params = Vec<(String, Type)>;
+// pub type Params = Vec<(String, Type)>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
```
Comment (Mossaka) [inline crates/wit-parser/src/lib.rs]:
```suggestion
#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
```

File: crates/wit-parser/src/lib.rs [hunk 5]
```diff
@@ -598,7 +636,7 @@ impl<'a> ExactSizeIterator for ResultsTypeIter<'a> {}
 impl Results {
     // For the common case of an empty results list.
     pub fn empty() -> Results {
-        Results::Named(Vec::new())
+        Results::Named(Params(Vec::new()))
```
Comment (Mossaka) [inline crates/wit-parser/src/lib.rs]:
```suggestion
        Results::Named(Vec::new().into())
```

File: crates/wit-parser/src/live.rs [hunk 6]
```diff
@@ -25,8 +25,8 @@ impl LiveTypes {
         }
     }
 
-    pub fn add_world(&mut self, resolve: &Resolve, world: WorldId) {
-        let world = &resolve.worlds[world];
+    pub fn add_world(&mut self, resolve: &Resolve, world_id: WorldId) {
```
Comment (Mossaka) [inline crates/wit-parser/src/live.rs]: No need to change?

**Review by Mossaka — APPROVED** [review]
lgtm!

File: crates/wit-parser/src/live.rs (hunk 6 above)
Comment (ydnar) [reply crates/wit-parser/src/live.rs]: Probably not, maybe a future PR?

**Review by Mossaka — APPROVED** [review]
LGTM! Thanks so much

**Pull request merged**